Perform PKeyAuth on NetCore and NetStandard

pmaytak opened the pull request:
Issue # 2363

- [x] Enable debugging project on a Win8 test machine
- [x] Update Microsoft.Identity.Test.Integration.Win8 with netcore and net5 targets.
- [x] NetCore: make code changes
- [x] NetCore: test code changes
- [x] Net5: make code changes
- [x] Net5: test code changes
- [x] NetStandard: make code changes
- [x] NetStandard test code changes
- [x] Refactor duplicated code in *DeviceAuthManager classes.

trwalke commented on src/client/Microsoft.Identity.Client/Platforms/netcore/NetCoreDeviceAuthManager.cs:
```diff
@@ -0,0 +1,61 @@
+                challengeData["SubmitUrl"] = endpointUri.AbsoluteUri;
+            }
+
+            try
```
lines 18 - 49 in this file is identical to what we have in the other device auth managers. Wouldnt hurt to move these to functions in the device auth helper.
For example, lines 38 to 49 can be "GetDeviceAuthCert()" and moved to the deviceAuthHelper. or perhaps you could use some kind of inheretence. Not too big of a deal but would further remove code duplication

pmaytak replied:
Done.

trwalke commented:
Build seems broken but that is probably due to some ADO issue. 
@pmaytak can you confirm that the PKEY Auth test run successfully on .net core and .net 5 on the win 8 agent?

jennyf19 commented on src/client/Microsoft.Identity.Client/Platforms/netcore/NetCoreDeviceAuthManager.cs:
```diff
@@ -0,0 +1,61 @@
+            IDictionary<string, string> challengeData = DeviceAuthHelper.ParseChallengeData(responseHeaders);
+
+            if (!challengeData.ContainsKey("SubmitUrl"))
+            {
```
nit: TryGetValue results in a single dictionary lookup, versus two w/ContainsKey and subsequent access.

pmaytak replied:
Done.

jennyf19 commented on src/client/Microsoft.Identity.Client/Platforms/netcore/NetCoreDeviceAuthManager.cs:
```diff
@@ -0,0 +1,61 @@
+                }
+            }
+
+            DeviceAuthJWTResponse responseJWT = new DeviceAuthJWTResponse(challengeData["SubmitUrl"],
```
>SubmitUrl [](start = 89, length = 9)

nit: use a variable

pmaytak replied:
Done.

jennyf19 commented on src/client/Microsoft.Identity.Client/Platforms/netcore/NetCoreDeviceAuthManager.cs:
```diff
@@ -0,0 +1,61 @@
+            DeviceAuthJWTResponse responseJWT = new DeviceAuthJWTResponse(challengeData["SubmitUrl"],
+                challengeData["nonce"], Convert.ToBase64String(certificate.GetRawCertData()));
+
+            byte[] signedResponse = new NetCoreCryptographyManager().SignWithCertificate(responseJWT.GetResponseToSign(), certificate);
```
can `responseJWT` be null?

pmaytak replied:
No, don't think so.

jennyf19 commented on src/client/Microsoft.Identity.Client/Platforms/netstandard13/NetStandardDeviceAuthManager.cs:
```diff
@@ -0,0 +1,61 @@
+{
+    class NetStandardDeviceAuthManager : IDeviceAuthManager
+    {
+        public bool TryCreateDeviceAuthChallengeResponseAsync(HttpResponseHeaders responseHeaders, Uri endpointUri, out string responseHeader)
```
+1 to what @trwalke said w/the duplicate code

pmaytak replied:
Done.

jennyf19 commented on src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DeviceAuthHelper.cs:
```diff
@@ -77,11 +75,99 @@ public static string GetBypassChallengeResponse(Dictionary<string, string> respo
+            string authHeaderTemplate = "PKeyAuth {0}, Context=\"{1}\", Version=\"{2}\"";
+
+            string signedJwt = string.Format(CultureInfo.InvariantCulture, "{0}.{1}", responseJWT.GetResponseToSign(), Base64UrlHelpers.Encode(signedResponse));
+            string authToken = string.Format(CultureInfo.InvariantCulture, " AuthToken=\"{0}\"", signedJwt);
```
question: can we do string interpolation here? it's like 5-10x faster than string.Format.

pmaytak replied:
Done.

jennyf19 commented on src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DeviceAuthHelper.cs:
```diff
@@ -77,11 +75,99 @@ public static string GetBypassChallengeResponse(Dictionary<string, string> respo
+
+            return signingCert[0];
+        }
     }
```
was this code added or moved from another file? wondering how closely to review it. thanks.

pmaytak replied:
@jennyf19 Yes, this was just extracted from the platform specific AuthManager classes.

jennyf19 replied:
ok cool. thanks. i won't review then..assume it's already been reviewed. thx

---
In reply to: [607430889](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2521#discussion_r607430889) [](ancestors = 607430889)

pmaytak commented:
> Build seems broken but that is probably due to some ADO issue.
> @pmaytak can you confirm that the PKEY Auth test runs successfully on .net core and .net 5 on the win 8 agent

Yea, I tested on Win 8 VM with VS and all targets (netcore, net5, netstandard) work.

bgavrilMS reviewed (approved):
LGTM, but agree with remaining comments about reducing code duplication if possible.

pmaytak commented:
Refactored: Added DeviceAuthManager class that has the common code and implements IDeviceAuthManager. Platform specific DeviceAuthManager do only getting the cert and signing (by overriding abstract methods).

jennyf19 commented on tests/Microsoft.Identity.Test.Integration.Win8/Microsoft.Identity.Test.Integration.Win8.csproj:
```diff
@@ -1,84 +1,50 @@
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.8.3" />
+    <PackageReference Include="MSTest.TestAdapter" Version="2.1.2" />
+    <PackageReference Include="MSTest.TestFramework" Version="2.1.2" />
+    <PackageReference Include="System.IdentityModel.Tokens.Jwt" Version="5.4.0" />
```
is this needed?

pmaytak replied:
No. Doesn't seem like it. Removed.

trwalke reviewed (approved):
LGTM

jennyf19 commented on src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DeviceAuthManager.cs:
```diff
@@ -0,0 +1,146 @@
+                X509Certificate2Collection signingCert = null;
+                signingCert = certCollection.Find(X509FindType.FindByThumbprint, challengeData["CertThumbprint"],
+                    false);
+                if (signingCert.Count == 0)
```
can `signingCert` be null?

pmaytak replied:
So, it can't be null here because Find returns at least empty collection. Source for [X509Certificate2Collection.Find](https://referencesource.microsoft.com/#System/security/system/security/cryptography/x509/x509certificate2collection.cs,174) and [X509Utils.GetCertificates](https://referencesource.microsoft.com/#System/security/system/security/cryptography/x509/x509utils.cs,522). But good point, had to look it up. :)

jennyf19 replied:
okay. thanks for checking. and +1 to combining those lines, if you don't mind.

jennyf19 reviewed (approved):
:shipit: 
Had one more question, but looks good. thx.

jennyf19 commented on src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DeviceAuthManager.cs:
```diff
@@ -0,0 +1,146 @@
+                }
+            }
+
+            if (signingCert == null || signingCert.Count == 0)
```
this should probably be what we do above, on line 92...maybe?

pmaytak replied:
This one can be null if cert authorities collection is empty and foreach never executes.

jennyf19 commented on src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DeviceAuthManager.cs:
```diff
@@ -0,0 +1,146 @@
+
+                signingCert = certCollection.Find(X509FindType.FindByIssuerDistinguishedName,
+                    distinguishedIssuerName, false);
+                if (signingCert.Count > 0)
```
i know you didn't write this, but this seems odd...and then check later if it's null...wouldn't it already throw here?...but maybe we don't want to change this now and just open an issue?

pmaytak replied:
Same as above, Find returns at least an empty collection.
